src/org/traccar/api/BaseObjectResource.java

```diff
@@ -0,0 +1,126 @@
+
+    private Class<T> baseClass;
+
+    public BaseObjectResource(Class<T> baseClass) {
+        this.baseClass = baseClass;
+    }
+
+    @POST
+    public Response add(T entity) throws SQLException {
+        Context.getPermissionsManager().checkReadonly(getUserId());
+        if (baseClass.equals(Device.class)) {
+            Context.getPermissionsManager().checkDeviceReadonly(getUserId());
+            Context.getPermissionsManager().checkDeviceLimit(getUserId());
+        }
+
+        BaseObjectManager<T> manager = Context.getManager(baseClass);
+        manager.addItem(entity);
+
+        Context.getDataManager().linkObject(User.class, getUserId(), baseClass, entity.getId(), true);
+
+        if (manager instanceof SimpleObjectManager) {
+            ((SimpleObjectManager<T>) manager).refreshUserItems();
+        }
+        if (baseClass.equals(Group.class) || baseClass.equals(Device.class)) {
+            Context.getPermissionsManager().refreshDeviceAndGroupPermissions();
+            Context.getPermissionsManager().refreshAllExtendedPermissions();
+        }
+        return Response.ok(entity).build();
+    }
+
+    @Path("{id}")
+    @PUT
+    public Response update(T entity) throws SQLException {
+        Context.getPermissionsManager().checkReadonly(getUserId());
+        if (baseClass.equals(Device.class)) {
+            Context.getPermissionsManager().checkDeviceReadonly(getUserId());
+        }
+        if (baseClass.equals(User.class)) {
+            User before = Context.getPermissionsManager().getUser(entity.getId());
+            Context.getPermissionsManager().checkUserUpdate(getUserId(), before, (User) entity);
+        }
+        Context.getPermissionsManager().checkPermission(baseClass, getUserId(), entity.getId());
+
+        Context.getManager(baseClass).updateItem(entity);
+
+        if (baseClass.equals(Group.class) || baseClass.equals(Device.class)) {
+            Context.getPermissionsManager().refreshDeviceAndGroupPermissions();
+            Context.getPermissionsManager().refreshAllExtendedPermissions();
+        }
+        if (baseClass.equals(User.class) && Context.getNotificationManager() != null) {
+            Context.getNotificationManager().refresh();
+        }
+        return Response.ok(entity).build();
+    }
+
+    @Path("{id}")
+    @DELETE
+    public Response remove(@PathParam("id") long id) throws SQLException {
+        Context.getPermissionsManager().checkReadonly(getUserId());
+        if (baseClass.equals(Device.class)) {
+            Context.getPermissionsManager().checkDeviceReadonly(getUserId());
+        }
+        Context.getPermissionsManager().checkPermission(baseClass, getUserId(), id);
+
+        BaseObjectManager<T> manager = Context.getManager(baseClass);
+        manager.removeItem(id);
+
+        if (manager instanceof SimpleObjectManager) {
+            ((SimpleObjectManager<T>) manager).refreshUserItems();
+        }
+        if (manager instanceof ExtendedObjectManager) {
```

Comment:
`ExtendedObjectManager` extends `SimpleObjectManager`, so it can be inside the first `if`.